src/Aspire.Hosting/ApplicationModel/ExpressionResolver.cs

```diff
@@ -0,0 +1,128 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Globalization;
+
+namespace Aspire.Hosting.ApplicationModel;
+
+internal class ExpressionResolver(string containerHostName, CancellationToken cancellationToken)
+{
+    class HostAndPortPresence
+    {
+        public bool HasHost { get; set; }
+        public bool HasPort { get; set; }
+    }
+
+    // For each endpoint, keep track of whether host and port are in use
+    // The key is the unique name of the endpoint, which is the resource name and endpoint name
+    readonly Dictionary<string, HostAndPortPresence> _endpointUsage = [];
+    static string EndpointUniqueName(EndpointReference endpointReference) => $"{endpointReference.Resource.Name}/{endpointReference.EndpointName}";
+
+    // This marks whether we are in the preprocess phase or not
+    // Not thread-safe, but we doesn't matter, since this class is never used concurrently
+    bool Preprocess { get; set; }
+
+
+    async Task<string?> EvalEndpointAsync(EndpointReference endpointReference, EndpointProperty property)
+    {
+        // In the preprocess phase, our only goal is to determine if the host and port properties are both used
```

Comment:
nit: resolve the unique name once in the beginning of the method and use it everywhere.